File: Priority Queue/1962_RemoveStonesToMinimizeTotal.cs
// 1962. Remove Stones to Minimize the Total
// Tags: priority queue (heap)
//

public class Solution 
{
    public int MinStoneSum(int[] piles, int k) 
    {
        // Edge cases.
        if(k < 1)
            return -1;
        else if (piles == null || piles.Length == 0)
            return -1;

        PriorityQueue<int, int> pq = new(new IntMaxComparer());
        int result = 0;

        // Populate priority queue
        foreach(var i in piles)
        {
            pq.Enqueue(i, i);
        }

        // Perform operation on largest value and decrement k.
        double temp;
        int temp2;
        while(k > 0)
        {
            temp = pq.Dequeue();
            temp2 = (int)Math.Ceiling(temp / 2);
            pq.Enqueue(temp2, temp2);

            k--;
        }

        // Sum up queue
        while(pq.Count > 0)
            result += pq.Dequeue();

       return result;
    }

    public class IntMaxComparer : IComparer<int>
    {
        public int Compare(int x, int y) => y.CompareTo(x);
    }
}

File: Priority Queue/1962_RemoveStonesToMinimizeTotal.rs
// 1962. Remove Stones to Minimize the Total
// Tags: priority queue (heap)
//

use std::collections::BinaryHeap;

impl Solution {
    pub fn min_stone_sum(piles: Vec<i32>, k: i32) -> i32 {
        
        // Edge cases.
        if piles.is_empty() {
            return -1;
        }
        else if k < 1 {
            return -1;
        }

        // Note: the binary_heap struct in Rust is implemented as 
        // a Max Heap.
        let mut pq = BinaryHeap::<i32>::with_capacity(piles.len());

        // Populate priority queue
        for el in piles {
            pq.push(el);
        }

        let mut k = k;
        while k > 0 {

            if let Some(el) = pq.pop() {
                let temp = el - (el / 2);
                pq.push(temp);
            }
            k -= 1;
        }

        pq.iter().sum()
    }
}